File: packages/typescript/src/judges/bayesian.ts
import type { DebateTranscript } from "../debate/engine.ts";
import type { JudgeStrategy, Verdict } from "./base.ts";

export type PersonaPriors = Record<string, Record<string, number>>;

export class BayesianJudge implements JudgeStrategy {
  private priors: PersonaPriors;

  constructor(priors: PersonaPriors = {}) {
    this.priors = priors;
  }

  async judge(transcript: DebateTranscript, labels: string[]): Promise<Verdict> {
    const finalRound = transcript.rounds[transcript.rounds.length - 1] ?? [];
    if (finalRound.length === 0) {
      return {
        label: transcript.primaryResult.label,
        confidence: transcript.primaryResult.confidence,
        reasoning: "No persona responses available. Falling back to primary result.",
        wasEscalated: true,
        primaryResult: transcript.primaryResult,
        debateTranscript: transcript,
        judgeStrategy: "bayesian",
        totalDurationMs: transcript.durationMs,
        totalCostUsd: transcript.totalCostUsd,
      };
    }

    const posterior = new Map<string, number>();
    labels.forEach((label) => posterior.set(label, 1));

    for (const response of finalRound) {
      for (const label of labels) {
        const prior = this.priors[response.personaName]?.[label] ?? 1 / Math.max(1, labels.length);
        const likelihood = response.label === label ? Number(response.confidence) : 1 - Number(response.confidence);
        posterior.set(label, (posterior.get(label) ?? 1) * Math.max(1e-6, prior * likelihood));
      }
    }

    const sum = Array.from(posterior.values()).reduce((a, b) => a + b, 0) || 1;
    let winner = labels[0] ?? transcript.primaryResult.label;
    let winnerScore = -1;

    for (const [label, score] of posterior.entries()) {
      const normalized = score / sum;
      if (normalized > winnerScore) {
        winner = label;
        winnerScore = normalized;
      }
    }

    return {
      label: winner,
      confidence: winnerScore,
      reasoning: "Bayesian aggregation across persona responses.",
      wasEscalated: true,
      primaryResult: transcript.primaryResult,
      debateTranscript: transcript,
      judgeStrategy: "bayesian",
      totalDurationMs: transcript.durationMs,
      totalCostUsd: transcript.totalCostUsd,
    };
  }
}

File: packages/python/src/llm_jury/judges/bayesian.py
from __future__ import annotations

from collections import defaultdict

from llm_jury.debate.engine import DebateTranscript

from .base import JudgeStrategy, Verdict, _fallback_verdict


class BayesianJudge(JudgeStrategy):
    def __init__(self, persona_priors: dict[str, dict[str, float]] | None = None) -> None:
        self.persona_priors = persona_priors or {}

    async def judge(self, transcript: DebateTranscript, labels: list[str]) -> Verdict:
        if not transcript.rounds or not transcript.rounds[-1]:
            return _fallback_verdict(transcript, "bayesian")

        posterior: dict[str, float] = defaultdict(lambda: 1.0)
        for response in transcript.rounds[-1]:
            for label in labels:
                prior = self.persona_priors.get(response.persona_name, {}).get(label, 1.0 / max(1, len(labels)))
                likelihood = response.confidence if response.label == label else (1.0 - response.confidence)
                posterior[label] *= max(1e-6, prior * likelihood)

        total = sum(posterior.values()) or 1.0
        for label in list(posterior.keys()):
            posterior[label] /= total

        winner = max(posterior, key=lambda k: posterior[k])
        return Verdict(
            label=winner,
            confidence=float(posterior[winner]),
            reasoning="Bayesian aggregation across persona responses.",
            was_escalated=True,
            primary_result=transcript.primary_result,
            debate_transcript=transcript,
            judge_strategy="bayesian",
            total_duration_ms=transcript.duration_ms,
            total_cost_usd=transcript.total_cost_usd,
        )
